### ISSUE TITLE
Problem with create table with custom type column

### ISSUE THREAD

**Issue by slad08** [fluentmigrator/fluentmigrator#1152]
Hello.My name is Denis. 
I've have some problem.I'll need create table via fluentmigrator.Below show sql script,for create this table.
I use SQL Server 2014, FluentMigrator v3.2 . 
In code  i also use  .AddSqlServer2014()
All information with scripts and code i attached 
[Problem with create table via fluentMigrator all scripts.txt](https://github.com/fluentmigrator/fluentmigrator/files/4077077/Problem.with.create.table.via.fluentMigrator.all.scripts.txt)
Could you help me?

P.S. 
If i deleted word NULL manually in script , what generated via fluentMigrator, when i describe customer types(in attache item 3 i mean DurationSecs and StartTimeDays columns) and run in MSSQL2014 management studion , this scrtipt working good.
How can i point in code, what don't need generated word NULL or NOT NULL(i mean item 3 in attached columns DurationSecs and StartTimeDays).
Not work:
[DurationSecs]  AS (DATEDIFF(SECOND, [StartTime], [EndTime])) PERSISTED NULL, 
[StartTimeDays]  AS (DATEDIFF(DAY,'19900101', [StartTime])) NULL)
if deleted null, work
[DurationSecs]  AS (DATEDIFF(SECOND, [StartTime], [EndTime])) PERSISTED, 
[StartTimeDays]  AS (DATEDIFF(DAY,'19900101', [StartTime])) )

**Comment by jzabroski** [comment]
I think the issue might be with the `NULL` keyword, not the custom type column, per say.

That being said, as a code review, I suggest you not make computed columns that depend on your server's datefmt settings, as '19900101' does.  I'm surprised SQL Server allows this , but it will definitely fail if you add PERSISTED to the end of this.

@fubar-coder  Do we have any support for computed columns?  It looks like we don't.

@slad08  As a workaround, first create the table without the computed columns, and then use an extension method to add the computed columns to the table:

```c#
public static class ColumnExtensions {
  public static void ComputedColumn(this Migration up, string columnName, string tableName, string schemaName, string computedColumnExpression, bool persisted)
  {
    this.Execute.Sql($@"ALTER TABLE [{schemaName}].[{tableName}] ADD [{columnName}] AS ({computedColumnExpression}){persisted ? " PERSISTED;" : ";"}");
  }
}
```

**Comment by slad08** [comment]
Thank's @jzabroski for you help. You answer helped me.

**Comment by slad08** [comment]
@fubar-coder @jzabroski  I have suggestion how to decide this problem in future for another customers of you fluentmigrator library. When customer want to create table with column AsCustom(), you can add additional argument method like bool format name IsNullable: Nullable-true (need write Null in sql requests), NotNullable-false( need write NotNull in sql requests), and by default without this argument(nothing to write in sql request)

**Comment by jzabroski** [comment]
@slad08 Thanks Denis.  I rephrased the issue as "Support Computed Columns"

**Comment by slad08** [comment]
@jzabroski  Ok, thanks.

**Comment by jzabroski** [comment]
Hi Denis,

I think we should always specify nullability.  For that reason, I don't think it's a good reason to co-opt custom type to derive a nullability phrase.  What would probably make more sense is to explicitly set Nullable(null) and create a method `Nullable(bool? nullableState)` where if nullability is unknown, we dont do anything.  I'll think about it.

> [`SET ANSI_NULL_DFLT_ON`](https://docs.microsoft.com/en-us/sql/t-sql/statements/set-ansi-null-dflt-on-transact-sql?view=sql-server-ver15) Modifies the behavior of the session to override default nullability of new columns when the ANSI null default option for the database is false.

**Comment by jzabroski** [comment]
> I think necessity to write words NULL or NOT NULL depends of version SQL server.

It turns out this is actually a database engine setting, believe it or not.

**Comment by MangelMaxime** [comment]
Hello,

having the `NULL` or `NOT NULL` always generated can generates invalid SQL queries.

For example, this query is invalid:

```sql
CREATE TABLE [dbo].[AuthTokens]
(
    [Id] INT NOT NULL IDENTITY(0,1),
    [CreationDateUTC] DATETIME NOT NULL,
    [LifeTimeSeconds] INT NOT NULL,
    [IsRevoked] BIT NOT NULL,
    [IsValid] AS CAST(IIF((DATEADD(ss, [LifeTimeSeconds], [CreationDateUTC]) > GETUTCDATE() AND [IsRevoked] = 0), 1, 0) AS BIT) NOT NULL,
    CONSTRAINT [PK_AuthTokens] PRIMARY KEY ([Id])
)
```

Error:

```
Msg 8183, Level 16, State 1, Line 7
Only UNIQUE or PRIMARY KEY constraints can be created on computed columns, while CHECK, FOREIGN KEY, and NOT NULL constraints require that computed columns be persisted.
```

If I remove the `NOT NULL` instruction on line `[IsValid]` then the SQL becomes valid.

I tried using `SET ANSI_NULL_DFLT_ON` but it didn't seems to work for me. I am using SQL Edge server.

**Comment by jzabroski** [comment]
@MangelMaxime Hello, I honestly wasn't aware of Azure SQL Edge.

The correct workaround is discussed here: https://github.com/fluentmigrator/fluentmigrator/issues/1152#issuecomment-575708512

I actually ran into this myself yesterday. I should probably smooth this feature out soon, as it is a bit annoying.

**Comment by MangelMaxime** [comment]
@jzabroski I didn't try the code you linked because I thought it would still add the `NULL` or `NOT NULL`.

Thanks for the tips

**Comment by jzabroski** [comment]
Nope, it works, although I did fix a typo 3 hours ago in the code where I had a missing ( ) in the interpolated format string. - I think I wrote that example from my phone many years ago w/o compiling.

**Comment by Shane32** [comment]
Any update here?  I would like to use `AsCustom` for computed columns within MS SQL Server as well.  As others have noticed, the only valid SQL generated is for persisted non-null columns.  See spec:

https://learn.microsoft.com/en-us/sql/t-sql/statements/create-table-transact-sql?view=sql-server-ver16

![image](https://github.com/user-attachments/assets/11e16683-a9c1-4ca9-9b8b-0675dc600cc4)

Currently FluentMigrator always adds either NULL or NOT NULL.  If the column is not persisted, it will always fail with an error that only persisted columns can be non-null (yes, it says this even for a nullable column).  If the column is persisted, and NULL is specified, it still gives an error.  Only when the column is persisted and NOT NULL is specified can the column be created, either within the Create.Table syntax or the Create.Column syntax.

```sql
# invalid and doesn't work
ALTER TABLE Product ADD Available AS InStock - Sold NULL
ALTER TABLE Product ADD Available AS InStock - Sold NOT NULL
ALTER TABLE Product ADD Available AS InStock - Sold PERSISTED NULL
# valid but no way to write fluent code
ALTER TABLE Product ADD Available AS InStock - Sold
ALTER TABLE Product ADD Available AS InStock - Sold PERSISTED
# valid and possible to write fluent code
ALTER TABLE Product ADD Available AS InStock - Sold PERSISTED NOT NULL
```

Currently the only functional code with `AsCustom` is as follows:

```cs
Create.Column("Available").OnTable("Product").AsCustom("AS InStock - Sold PERSISTED").NotNullable();
```

Ideally the code would be more like this:

```cs
// ALTER TABLE Product ADD Available AS InStock - Sold
Create.Column("Available").OnTable("Product").AsComputed("InStock - Sold");
// ALTER TABLE Product ADD Available AS InStock - Sold PERSISTED
Create.Column("Available").OnTable("Product").AsComputed("InStock - Sold").Persisted();
// ALTER TABLE Product ADD Available AS InStock - Sold PERSISTED NOT NULL
Create.Column("Available").OnTable("Product").AsComputed("InStock - Sold").Persisted().NotNullable();
```

For now I am using `Execute.Sql` as a workaround.

**Comment by jzabroski** [comment]
I've exposed ITypeMap. It's a start. We would need a new AsExpression method. Expression input would then have it's own quoting rules.

**Comment by Shane32** [comment]
Summary by provider: (probably should be double checked)

| Database             | Computed Columns Supported? | Syntax |
|----------------------|-----------------------------|--------|
| **DB2**              | Yes                         | `<name> <dataType> GENERATED ALWAYS AS <expression> [ STORED ]` |
| **Firebird**         | Yes                         | `<name> <dataType> GENERATED ALWAYS AS <expression>` |
| **SAP HANA**         | Yes                         | `<name> <dataType> GENERATED ALWAYS AS <expression> [ STORED ]` |
| **Jet**              | No                          | N/A |
| **MySQL**            | Yes                         | `<name> <dataType> GENERATED ALWAYS AS <expression> { VIRTUAL \| STORED }` |
| **Oracle**           | Yes                         | `<name> <dataType> GENERATED ALWAYS AS <expression>` |
| **PostgreSQL**       | Only Persisted              | `<name> <dataType> GENERATED ALWAYS AS <expression> STORED` |
| **Amazon Redshift**  | No                          | N/A |
| **Snowflake**        | No                          | N/A |
| **SQLite**           | Yes                         | `<name> <dataType> GENERATED ALWAYS AS <expression> { VIRTUAL \| STORED }` |
| **SQL Server**       | Only Persisted              | `<name> AS (<expression>) [ PERSISTED ]` |

Syntax needs to be expanded to include nullable syntax.

**Comment by jzabroski** [comment]
That's beautiful. Do you want to try putting together an initial PR? I am thinking of dropping support for Jet, as it isn't supported in any .NET Core runtime versions. Db2 is and I am working on fixing the references to use the official IBM package.

**Comment by Shane32** [comment]
> Do you want to try putting together an initial PR?

Yes, I will do so if I have time.  I started already, and I think I see what needs to be done.

**Comment by Shane32** [comment]
@jzabroski Perhaps you can suggest the syntax for me to follow.  So far I've added `AsExpression(string expression)`, which does not clear `Type` so that you can use `.AsInt32().AsExpression("col1 + col2")` as the type is necessary for everything except SQL Server.  But what about virtual/stored aka computed/persisted?  Options:

1. Add another parameter to `AsExpression` such as `AsExpression(string expression, bool stored = false)`
2. Design another interface that `AsExpression` will return so that you can call `.AsExpression("col1 + col2").Stored()` and similar

Should virtual/computed be default?  I assume yes.

Naming -- `Virtual`/`Stored`?  `Computed`/`Persisted`?  Other?

Implementation details?  So far I added `string Expression { get; set; }` to `IColumnDataType` with other corresponding changes, plus another virtual method to `ColumnBase` called `FormatExpression`.  If breaking changes (by adding methods to interfaces) are not allowed, `GetAdditionalFeature` and similar could be used, but since it seems like the feature seems to be generally available in similar syntax across all databases (notably except SQL Server), I expect it would be better to add properties and methods where needed.  I also need to store virtual/stored somewhere.  Another prop - `bool ExpressionVirtual { get; set; }`?

If you can provide some guidance as to implementation by answering the above, I may be better able to post a PR - thanks!

**Comment by jzabroski** [comment]
I asked ChatGPT your question. Based on its notes, I would say an expression should default to not stored, and if it is stored, we save it as stored.  We can consider adding an extension method to SqlServerExtensions called Persisted() for people who would rather a native syntax but it seems like a waste to me.

I would add string Expression { get; set; } and bool ExpressionStored { get; set; } to IColumnDataType, yes. Up to you if you want to wrap the two into a single wrapper type called ExpressionType. I don't know if it makes the code cleaner - have not spent brain cycles thinking about it. You should decide what is pleasing to your eye.

> **PROMPT**
> You are a database engineer tasked with deciding the vocabulary for a new feature in the database engine.  The feature is to store a calculated result as a column.  Based on competitors, the following terms are used:  
`Virtual`/`Stored`? `Computed`/`Persisted`? Other recommendations?  Explain your reasoning.

> **ANSWER**
> In the context of storing a calculated result as a column in a database engine, the terms Virtual and Stored are commonly used to describe different approaches.
> 
> Virtual columns refer to columns that are not physically stored in the database, but rather calculated on the fly when requested. These columns are defined by an expression or formula based on other columns in the table. Since the values are calculated dynamically, they do not occupy additional storage space. However, they may impact performance as the calculation needs to be performed every time the column is accessed.
> 
> Stored columns, on the other hand, refer to columns that are physically stored in the database, just like regular columns. The values in these columns are calculated and stored when the data in the table is updated. This approach can improve performance as the calculated values are readily available without the need for recalculating them each time.
> 
> For the terms Computed and Persisted, they are often used interchangeably with Virtual and Stored, respectively. The choice between these terms largely depends on the terminology used in the existing database ecosystem or the preferences of the development team.
> 
> In addition to the mentioned terms, another recommendation could be to use Derived or Calculated. These terms convey the idea that the column's value is derived from other columns or calculated based on specific rules or formulas.
> 
> Ultimately, the choice of vocabulary depends on the specific context, the preferences of the development team, and the need for consistency with existing terminology in the database ecosystem.

**Comment by jzabroski** [comment]
> Add another parameter to AsExpression such as AsExpression(string expression, bool stored = false)

I think this is probably the more pleasurable way to type this for now.  I don't believe there will be a ton more settings on this such that the .Stored() "finally tagless syntax" adds a ton of value.

**Comment by PhenX** [comment]
Closed by #1896
